api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java

```diff
@@ -490,4 +514,93 @@ public FirecloudNihStatus getNihStatus() {
           }
         });
   }
+
+  @Override
+  public void addProjectToServicePerimeter(String servicePerimeterName, String billingProject) {
+    final String utf8 = StandardCharsets.UTF_8.name();
+
+    // yes this actually gets URL decoded twice
+    // TODO: update Rawls, then FC Orch, then AoU.  See Terra JIRA AS-559
+    final String doublyEncodedName;
+    try {
+      doublyEncodedName = URLEncoder.encode(URLEncoder.encode(servicePerimeterName, utf8), utf8);
+    } catch (UnsupportedEncodingException e) {
+      throw new ServerErrorException(e);
+    }
+
+    ServicePerimetersApi perimetersApi = servicePerimetersApiProvider.get();
+    retryHandler.run(
+        (context) -> {
+          try {
+            perimetersApi.addProjectToServicePerimeter(doublyEncodedName, billingProject);
+          } catch (ApiException e) {
+            String alreadyInMsg =
+                String.format(
+                    "project %s is already in service perimeter %s",
+                    billingProject, servicePerimeterName);
+            if (configProvider.get().featureFlags.badRequestIsOkWhenAlreadyInPerimeter
+                && e.getCode() == 400
+                && e.getResponseBody().contains(alreadyInMsg)) {
+              log.info(alreadyInMsg);
+            } else {
+              throw e;
+            }
+          }
+          return null;
+        });
+
+    // it takes some time to add the project to the perimeter
+    waitForReadyProject(billingProject);
+  }
+
+  private void waitForReadyProject(String billingProject) throws WorkbenchException {
+    log.info(String.format("Waiting for billing project %s to become READY", billingProject));
+    try {
+      final CreationStatusEnum status =
+          terminalStatusRetryer()
+              .call(() -> getBillingProjectStatus(billingProject))
+              .getCreationStatus();
+
+      if (status == CreationStatusEnum.READY) {
+        log.info(String.format("Billing project %s is READY", billingProject));
+      } else {
+        throw new WorkbenchException(
+            String.format("Billing project %s has %s status", billingProject, status.getValue()));
+      }
+    } catch (RetryException | ExecutionException e) {
+      throw new WorkbenchException(
+          String.format(
+              "Timed out waiting for billing project %s to transition to READY status after %d seconds",
+              billingProject, configProvider.get().firecloud.timeoutInSeconds),
+          e);
+    }
+  }
+
+  // I'd love to use our existing Spring retry system but this does not seem to be possible
```

Comment:
unless I'm missing a more-Spring-native alternative here?

Reply:
I looked into that.  The RetryContext does not seem to have any reference to returned values - only exceptions.